### PR TITLE
Adds prettier as a recommended extension w/ format on save

### PR DESCRIPTION
**Pull request by jlsnow301**
## About The Pull Request
- Adds [prettierx](https://github.com/brodybits/prettierx) as a dependency (aliased as prettier)
- Adds prettier as a recommended extension
- Sets format on save as default for JS/TS files
- Validates the usage of prettier in CI
- Disables many conflicting rules within ESLint
- Adds [typescript-eslint plugin](https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/eslint-plugin) for typescript no-unused-vars support
- Updates ESLint and its other packages
- Refactors some low hanging fruit into TypeScript
- Refactors some older/poor conventions by hand.

## Adding this to your project:
<details>
<summary>View</summary>

Base assumptions: You have a nodejs project with yarn installed.

1. Change to your working folder, in this case, `cd tgui` 
2. `yarn add prettier@npm:prettierx@0.19.0` (OR LATEST VERSION)
3. `yarn dlx @yarnpkg/sdks vscode`
4. set prettier's location in .vscode/settings.json `"prettier.prettierPath": "./tgui/.yarn/sdks/prettier/index.js",`
5. add prettier as a recommended extension in .vscode/extensions.json: `"esbenp.prettier-vscode"`
6. set prettier as the default formatter for each filetype you want formatted in settings: `"editor.defaultFormatter": "esbenp.prettier-vscode",`
7. set format on save in the same filetype: `"editor.formatOnSave": true`
8. If you want a CLI check, add this package.json: `script-name: prettierx --check folder/to/check`
9. You can format packages in subfolders by running `yarn prettier --write folder/to/format`
10. If you are using eslint, you will need to disable many rules associated with indentation etc. Prettier now handles these, so they need to be configured to play nice with each other: [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) or manually.

</details>

## Why It's Good For The Game
Enforces consistency and therefore readability of the (interface portion of) code

Prettier also saves developer time as it corrects syntactical errors on its own, letting them focus on the objective.

## Changelog

:cl:
code: Prettier is now recommended for UI development. 
refactor: Many interfaces have been formatted and/or refactored, report bugs you see in Issues.
/:cl:

### CODE REVIEW

**Comment by jlsnow301** [comment]
working! check linters

**Comment by MrStonedOne** [comment]
My feelings on this are basically gonna depend on what the prettier rules are.

which should be disclosed in this pr or if already documented, at least linked to.

**Comment by jlsnow301** [comment]
> My feelings on this are basically gonna depend on what the prettier rules are.

style created the current prettier rules file which are as follows
`arrowParens: always
bracketSpacing: true
endOfLine: lf
jsxBracketSameLine: true
jsxSingleQuote: false
printWidth: 80
proseWrap: preserve
quoteProps: preserve
semi: true
singleQuote: true
tabWidth: 2
trailingComma: es5`

**Comment by Mothblocks** [comment]
that should be the same prettier that style set up a bit ago. the most controversial of those is probably the line width, which is already enforced by CI

**Comment by stylemistake** [comment]
Prettierx, one line imports are required

https://github.com/prettier/prettier/issues/5995#issuecomment-1101771951

**Comment by jlsnow301** [comment]
> Prettierx, one line imports are required

Is this just prettier with more options? This is a nice find

**Comment by Mothblocks** [comment]
Will that work with someone freshly cloning the repository?

**Comment by stylemistake** [comment]
Yes, as long as you alias it properly in yarn, it will appear as if it's the real prettier.

**Comment by Mothblocks** [comment]
"You" as in us with the comment you provided, yeah? If so then works for me 👍

File: tgui/package.json [hunk 1]
```diff
@@ -11,6 +11,7 @@
     "tgui:analyze": "webpack --analyze",
     "tgui:dev": "node --experimental-modules packages/tgui-dev-server/index.js",
     "tgui:lint": "eslint packages --ext .js,.cjs,.ts,.tsx",
+    "tgui:prettier": "prettierx --check 'packages/tgui/interfaces/**/*.{js,jsx,ts,tsx}'",
```
Comment (stylemistake) [inline tgui/package.json:14]: This should work.

```suggestion
    "tgui:prettier": "prettier --check 'packages/tgui/interfaces/**/*.{js,jsx,ts,tsx}'",
```

Comment (jlsnow301) [reply tgui/package.json:14]: resolved in discord

File: tgui/.prettierrc.yml [hunk 2]
```diff
@@ -1,8 +1,9 @@
 arrowParens: always
-bracketSpacing: true
```
Comment (stylemistake) [inline tgui/.prettierrc.yml:2]: Why did you remove bracket spacing?

Comment (jlsnow301) [reply tgui/.prettierrc.yml:2]: This does not exist on prettierx apparently, or at least is categorized into subtypes

Comment (stylemistake) [reply tgui/.prettierrc.yml:2]: i'd still keep it

**Comment by MrStonedOne** [comment]
why are js files not tab indented?

We are a tab indented project.

File: tgui/.prettierrc.yml (hunk 2 above)
Comment (jlsnow301) [reply tgui/.prettierrc.yml:2]: It throws a warning on ci, can we  use one of its granular subtypes?

**Comment by Mothblocks** [comment]
did you mess it up with a force push again

**Comment by jlsnow301** [comment]
> did you mess it up with a force push again

seems like this whole pr is about messing things up

**Comment by jlsnow301** [comment]
> We are a tab indented project.

prettier now automatically fixes this!

File: tgui/.prettierrc.yml [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
 endOfLine: lf
+importFormatting: oneline
 jsxBracketSameLine: true
 jsxSingleQuote: false
-printWidth: 80
+printWidth: 72
```
Comment (jlsnow301) [inline tgui/.prettierrc.yml:6]: you might wonder why this is set to 72. printwidth in prettier is not like eslint, it's not a "rule" but a suggestion. having a printwidth of 80 will still create lines that go over that figure, so i'm tuning it down to 72.

Comment (stylemistake) [reply tgui/.prettierrc.yml:8]: eeeh, i think 80 will actually work better, more is always better with typescript

**Review by MrsTonedOne — APPROVED** [review]

File: tgui/.eslintrc.yml [hunk 4]
```diff
@@ -482,11 +488,8 @@ rules:
   ## Enforce consistent spacing before blocks
   space-before-blocks: [error, always]
   ## Enforce consistent spacing before function definition opening parenthesis
-  space-before-function-paren: [error, {
-    anonymous: always,
-    named: never,
-    asyncArrow: always,
-  }]
+  space-before-function-paren:
+    [error, { anonymous: always, named: never, asyncArrow: always }]
```
Comment (stylemistake) [inline tgui/.eslintrc.yml:492]: can it not format eslintrc please

File: tgui/.prettierrc.yml [hunk 5]
```diff
@@ -1,12 +1,15 @@
...
-printWidth: 80
+offsetTernaryExpressions: true
+printWidth: 72
 proseWrap: preserve
 quoteProps: preserve
 semi: true
 singleQuote: true
 tabWidth: 2
 trailingComma: es5
+useTabs: true
```
Comment (stylemistake) [inline tgui/.prettierrc.yml:15]: This would suck for code reviews on Github, because Github treats tabs as 8 spaces regardless of our editorconfig, and a lot of auto-generated javascript and existing javascript on the web (and here) already uses two spaces for tabs.

File: tgui/.eslintrc.yml [hunk 6]
```diff
@@ -708,11 +711,9 @@ rules:
   ## Enforce event handler naming conventions in JSX
   react/jsx-handler-names: error
   ## Validate JSX indentation (fixable)
-  react/jsx-indent: [error, 2, {
-    checkAttributes: true,
-  }]
+  # react/jsx-indent: [error, tab, { checkAttributes: true }]
   ## Validate props indentation in JSX (fixable)
-  react/jsx-indent-props: [error, 2]
+  react/jsx-indent-props: [error, tab]
```
Comment (stylemistake) [inline tgui/.eslintrc.yml:716]: Disable all indent rules because now prettier handles it, less rules for eslint means it finishes linting wuicker.

File: tools/build/build.js [hunk 7]
```diff
@@ -194,6 +194,10 @@ export const TguiEslintTarget = new Juke.Target({
   dependsOn: [YarnTarget],
   executes: ({ get }) => yarn('tgui:lint', !get(CiParameter) && '--fix'),
 });
+export const TguiPrettierTarget = new Juke.Target({
```
Comment (stylemistake) [inline tools/build/build.js:197]:
```suggestion

export const TguiPrettierTarget = new Juke.Target({
```

File: tgui/.prettierrc.yml (hunk 5 above)
Comment (Mothblocks) [reply tgui/.prettierrc.yml:15]: tbf the existing javascript would ctrl+s into using tabs if you have vsc setup, and it is odd for us to have tabs in one end of the codebase and spaces in another for reasons not related to something like YML requiring spaces or similar

File: tgui/.eslintrc.yml (hunk 6 above)
Comment (jlsnow301) [reply tgui/.eslintrc.yml:716]: It does, and I can, but I was thinking we should leave some warnings on to let players know they're coding against standards (if they don't have prettier).
I can revert the previous commit for this

File: tgui/.prettierrc.yml (hunk 3 above)
Comment (jlsnow301) [reply tgui/.prettierrc.yml:8]: it will make lines more than 80, so it's your call
people really need to start breaking their components down so we aren't even reaching that point

Comment (jlsnow301) [reply tgui/.prettierrc.yml:8]: you should try this locally and tell me what you think. it's seems to be doing well, i've been testing it

Comment (stylemistake) [reply tgui/.prettierrc.yml:8]: > having a printwidth of 80 will still create lines that go over that figure

only if it can't fit something under 80, if it was 72, it would have the same issue, but it would more aggressively newline everything which is undesirable.

Comment (jlsnow301) [reply tgui/.prettierrc.yml:8]: it has been reverted to 80 for the coming commit

File: tgui/packages/tgui-say/handlers/arrowKeys.tsx [hunk 8]
```diff
@@ -1,14 +1,13 @@
...
 /** Increments the chat history counter, looping through entries */
-export const handleArrowKeys = function (this: Modal, direction: number) {
+export const handleArrowKeys = function (this: Modal, direction: string) {
 	const { historyCounter } = this.fields;
-	if (direction === KEY_UP && historyCounter < getHistoryLength()) {
+	if (direction === 'ArrowUp' && historyCounter < getHistoryLength()) {
 		this.fields.historyCounter++;
 		this.events.onViewHistory();
-	} else if (direction === KEY_DOWN && historyCounter > 0) {
+	} else if (direction === 'ArrowDown' && historyCounter > 0) {
```
Comment (Mothblocks) [inline tgui/packages/tgui-say/handlers/arrowKeys.tsx]: Why?

Comment (jlsnow301) [reply tgui/packages/tgui-say/handlers/arrowKeys.tsx]: keycode is deprecated and i'm tired of looking at the error
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

Comment (stylemistake) [reply tgui/packages/tgui-say/handlers/arrowKeys.tsx]: What error?

File: tgui/packages/tgui/interfaces/NtosCrewManifest.js [hunk 9]
```diff
@@ -4,8 +4,8 @@ import { Button, Section, Table } from '../components';
 import { NtosWindow } from '../layouts';
 
 export const NtosCrewManifest = (_, context) => {
-	const { act, data } = useBackend(context);
-	const { have_printer, manifest = {} } = data;
+	const { act, data } = useBackend < Data > context;
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/NtosCrewManifest.js:7]: what

Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/NtosCrewManifest.js:7]: this isnt tsx friend

**Comment by jlsnow301** [comment]
Project is getting a little out of hand, going to atomize it

File: tgui/packages/tgui/interfaces/NtosCrewManifest.js (hunk 9 above)
Comment (jlsnow301) [reply tgui/packages/tgui/interfaces/NtosCrewManifest.js:7]: walt gasp.jpg
